### PR TITLE
User/bena/terminology load

### PR DESCRIPTION
**Pull request by petermcclanahan**
Changes to make MDS and DPS calls sequential

### CODE REVIEW

File: Scripts/Install-Terminology.ps1 [hunk 1]
```diff
@@ -32,13 +36,15 @@
...
+    [String] $LoaderWindowsUser,
+    [String] $ProcessingServiceWindowsUser,
     [switch] $Quiet
 )
 
 # Fail installation on first error
 $ErrorActionPreference = "Stop"
 
-[string[]] $requiredFiles = @($InstallFile, $Dacpac, $PublishProfile, "$PSScriptRoot\Install.config", "$PSScriptRoot\Terminology-Install-Utilities.psm1", "$PSScriptRoot\terminology-registration.config")
+[string[]] $requiredFiles = @($InstallFile, $Dacpac, $PublishProfile, "$PSScriptRoot\Install.config", "$PSScriptRoot\terminology-registration.config", "$PSScriptRoot\Terminology.json", "$PSScriptRoot\SharedTerminology.json", "$PSScriptRoot\Terminology-Install-Utilities.psm1")
```
Comment (petermcclanahan) [inline Scripts/Install-Terminology.ps1]: Nice

File: Scripts/Terminology-Install-Utilities.psm1 [hunk 2]
```diff
@@ -4,7 +4,12 @@ function Get-UserValueOrDefault {
         [String] $Prompt
     )
 
-    $value = Read-Host "Enter the $Prompt or hit enter to accept the default value [$Default]"
+    Write-Host "Enter the " -NoNewline -ForegroundColor White
+    Write-Host $Prompt  -NoNewline -ForegroundColor Yellow
```
Comment (nickrflynn) [inline Scripts/Terminology-Install-Utilities.psm1]: Nice touch

**Comment by petermcclanahan** [comment]
I was still seeing some Fabric errors. Did you get those resolved?

image

File: Scripts/Terminology-Install-Utilities.psm1 [hunk 3]
```diff
@@ -355,6 +408,12 @@ function Get-TerminologyConfig {
     }
     $sqlLogDirectoryConfig = Get-ConfigValue -Prompt "Log directory to create Terminology database" -DefaultFromParam $SqlLogDirectory -DefaultFromInstallConfig $sqlLogDirectoryParam -Quiet:$Quiet
 
+    # Loader User
+    $loaderWindowsUser = Get-ConfigValue -Prompt "Loader windows username" -AdditionalPromptInfo "(ex. DOMAIN\USER)" -DefaultFromParam $LoaderWindowsUserParam -DefaultFromInstallConfig $installSettings.loaderWindowsUser -Quiet:$Quiet
```
Comment (nickrflynn) [inline Scripts/Terminology-Install-Utilities.psm1]: Small thing but "windows" should probably be capitalized

File: Scripts/Terminology-Install-Utilities.psm1 [hunk 4]
```diff
@@ -520,29 +576,65 @@ function Invoke-PostToMds {
...
     $headers = @{"Content-Type" = "application/json"}
-    $headers.Add("Authorization", "Bearer $authToken")
+    $headers.Add("Authorization", "Bearer $($Config.accessToken)")
 
     try {
-        Write-DosMessage -Level "Information" -Message "Starting to create metadata for $name"
-        $response = Invoke-RestMethod -Uri "$($Config.mdsServiceUrl)/DataMarts" -Method POST -Body $dataMartJson -Headers $headers -UseBasicParsing
-        Write-DosMessage -Level "Information" -Message "Completed creating metadata for $name"
-        return $response.Id
+        Write-DosMessage -Level "Information" -Message "Posting json metadata in $path for $name in mds $($Config.mdsServiceUrl)/DataMarts.`nThis may take several minutes"
```
Comment (nickrflynn) [inline Scripts/Terminology-Install-Utilities.psm1]: We should capitalize MDS (and possibly JSON)

File: Scripts/Terminology-Install-Utilities.psm1 [hunk 5]
```diff
@@ -520,29 +576,65 @@ function Invoke-PostToMds {
...
+        $response = Invoke-RestMethod -Uri "$($Config.mdsServiceUrl)/DataMarts" -Method POST -Body $dataMartJson -Headers $headers -TimeoutSec 600 -UseBasicParsing
+        Write-DosMessage -Level "Information" -Message "Completed creating metadata for $name with data mart $($response.Id)"
+        return $($response.Id)
     }
     catch [System.Net.WebException] {
-        Write-DosMessage -Level "Error" -Message "Creating metadata for $name stopped" -ErrorAction Continue
-        Write-DosMessage -Level "Error" -Message "Description:" $_.Exception.Response.StatusDescription -ErrorAction Continue
-        Write-DosMessage -Level "Error" -Message  $_.ErrorDetails.Message | ConvertFrom-Json | Select-Object -Expand message -ErrorAction Continue
-        throw $_.Exception
+        #Write-DosMessage -Level "Error" -Message "Creating metadata for $name stopped"
```
Comment (nickrflynn) [inline Scripts/Terminology-Install-Utilities.psm1]: Why are these included and commented out?

File: Scripts/Terminology-Install-Utilities.psm1 [hunk 6]
```diff
@@ -520,29 +576,65 @@ function Invoke-PostToMds {
...
+        #Write-DosMessage -Level "Error" -Message "Description:" $_.Exception.Response.StatusDescription
+        #Write-DosMessage -Level "Error" -Message  $_.ErrorDetails.Message | ConvertFrom-Json | Select-Object -Expand message
+        Write-Host $_.ErrorDetails.Message | ConvertFrom-Json | Select-Object -Expand message
+        Write-Host "Second"
```
Comment (nickrflynn) [inline Scripts/Terminology-Install-Utilities.psm1]: ??

File: Scripts/Terminology-Install-Utilities.psm1 [hunk 7]
```diff
@@ -551,61 +643,68 @@ function Invoke-PostToDps {
...
+        Write-DosMessage -Level "Information" -Message  "Batch execution successfully sent to the data processing service."
 
         return $batchExecutionId;
     }
     catch [System.Net.WebException] {
-        Write-DosMessage -Level "Error" -Message  "POST for '$name' failed with status code:" $_.Exception.Response.StatusCode.value__ -ErrorAction Continue
-        Write-DosMessage -Level "Error" -Message  "Description:" $_.Exception.Response.StatusDescription -ErrorAction Continue
-        throw $_.Exception
+        Write-Host $_.ErrorDetails.Message | ConvertFrom-Json | Select-Object -Expand message
+        Write-Host "Second"
```
Comment (nickrflynn) [inline Scripts/Terminology-Install-Utilities.psm1]: ?? (again)

File: Scripts/Terminology-Install-Utilities.psm1 [hunk 8]
```diff
@@ -551,61 +643,68 @@ function Invoke-PostToDps {
...
+        Write-Host $_.Exception.Message
+        Write-Host $_.Exception.Response
+        Write-Host $_.Exception.InnerException
+        Write-Host $_.Exception.InnerException | ConvertFrom-Json | Select-Object -Expand innerException
+        Write-Host $_.Exception | ConvertFrom-Json
+        Write-Host $_.Exception | ConvertFrom-Json | Select-Object -Expand message
+        Write-Host ($_.Exception | Format-Table | Out-String)
+        
+        #Write-DosMessage -Level "Error" -Message  "POST for '$name' failed with status code:" $_.Exception.Response.StatusCode.value__
```
Comment (nickrflynn) [inline Scripts/Terminology-Install-Utilities.psm1]: Commented code?  Also, I believe we want to rethrow the exception.

File: Scripts/Terminology-Install-Utilities.psm1 (hunk 4 above)
Comment (nickrflynn) [reply Scripts/Terminology-Install-Utilities.psm1]: Or maybe expand it to "metadata service"

File: Scripts/Terminology-Install-Utilities.psm1 (hunk 3 above)
Comment (benjanderson) [reply Scripts/Terminology-Install-Utilities.psm1]: 👍

File: Scripts/Terminology-Install-Utilities.psm1 (hunk 4 above)
Comment (benjanderson) [reply Scripts/Terminology-Install-Utilities.psm1]: 👍

File: Scripts/Terminology-Install-Utilities.psm1 [hunk 9]
```diff
@@ -698,80 +797,163 @@ function Add-MetadataAndStructures() {
...
-    # POST executions 
-    Write-DosMessage -Level "Information" -Message "Creating physical tables for Terminology and SharedTerminology data marts"
-    $terminologyBatchExecutionId = Invoke-PostToDps -Config $Config -dataMartId $terminologyDataMartId
-    $sharedTerminologyBatchExecutionId = Invoke-PostToDps -Config $Config -dataMartId $sharedTerminologyDataMartId
+    $terminologyExists = Test-DataMartExists -Name "Terminology" -Config $Config
+    $sharedTerminologyExists = Test-DataMartExists -Name "SharedTerminology" -Config $Config
 
-    # Poll batch executions for 5 minutes to determine if they've been successful
-    $wasSuccessful = Invoke-PollBatchExecutions -Config $Config -terminologyBatchExecutionId $terminologyBatchExecutionId -sharedTerminologyBatchExecutionId $sharedTerminologyBatchExecutionId
+    if (($terminologyExists -eq $false) -and ($sharedTerminologyExists -eq $false)) {
```
Comment (petermcclanahan) [inline Scripts/Terminology-Install-Utilities.psm1]: This section looks like a bad merge (probably the one I did)...

File: Scripts/Terminology-Install-Utilities.psm1 [hunk 10]
```diff
@@ -698,80 +797,163 @@ function Add-MetadataAndStructures() {
...
+        if(!$terminologyDataMartId) {
```
Comment (petermcclanahan) [inline Scripts/Terminology-Install-Utilities.psm1]: Lines 806-814 should be removed

Comment (benjanderson) [reply Scripts/Terminology-Install-Utilities.psm1]: yup, bad merge. I think I have resolved it and am fixing it

File: Scripts/Terminology-Install-Utilities.psm1 (hunk 9 above)
Comment (benjanderson) [reply Scripts/Terminology-Install-Utilities.psm1]: 👍

**Comment by benjanderson** [comment]
https://github.com/HealthCatalyst/Fabric.Terminology/pull/163#issuecomment-422464988
He didn't have the Shared db. We resolved it
